src/flask_pyoidc/flask_pyoidc.py

```diff
@@ -203,12 +206,20 @@ def _show_error_response(self, error_response):
 
         return 'Something went wrong with the authentication, please try to login again.'
 
-    def oidc_auth(self, provider_name):
+    def oidc_auth(self, provider_name, accept_token: bool = None,
```

Comment:
I think this should be a separate decorator function to avoid mixing authorization methods.
Either an endpoint should rely on the user session stored in cookies, or it should accept access tokens.

Reply:
Initially, I implemented that design but the problem I faced with that was I was not able to test my endpoints with Swagger UI. I use falsk-restx to document APIs for flask app.

Within flask-restx, Swagger can be configured to use Oauth2 Authorization which sends token in the authorization header of the request. But because the endpoint is expecting cookie, it sends redirect response to Swagger instead of executing the view function.

Also, if you have a flask web app that has a developer feature to issue tokens to access your APIs, that won't work if APIs are made to work with either cookie only or token only. E.g. From Github developer settings, you can generate token to access Github APIs from your code that you normally do from Github web app.

Then I came up with the second approach for "why not expect both?" and go ahead with what is available. So if the user has valid cookie, access will be given, if not cookie, atleast expect token else if there's nothing, redirect the user.

I will push the demo app that I've created for this pull request once I'm done commenting it.

Reply:
I don't fully understand what issue you ran into with Swagger UI - it should support regular Authorization flow (though I do know there might be some issues with CORS).

But actually in terms of Github (and generally good API design) you shouldn't mix cookies and tokens as authorization.
As far as I know there's no public API that relies on cookies, instead you get a token (e.g. via Github developer settings) which you then use with defined REST APIs. Those same API endpoints do not accept cookies.

Typically tokens are used for authorization when there is no active user session, i.e. no cookies.
So I still think it's better to completely separate the concerns of auth via token or cookie.
That will help prevent accidental misuse.

Reply:
In order to separate it, I need to address one condition. What if there's a private API that is meant to be accessible for Alice who is accessing it through the web app in browser as well as for Bob who wants to access the API through curl or REST API client with access_token?

Should we allow to decorate the endpoint twice (one for OIDC & one for access_token)? I want to make it look seamless and with minimal changes to the user codebase. Something that can be turned on & off with an argument, per endpoint.

Reply:
I don't expect that to be a very common use case, and the extra complexity is not worth it IMO when the user codebase work would be minimal. I imagine it actually being two separate endpoints that share the majority of code, something like:

```python
@app.route('/api/doWork')
@auth.token_auth(PROVIDER_NAME1)
def apiDoWork():
    authData = <use/extract token data>
    doWork(authData)

@app.route('/web/doWork')
@auth.oidc_auth(PROVIDER_NAME1)
def webDoWork():
    userSession = UserSession(flask.session)
    doWork(userSession)
```

Reply:
Code duplication will deter developers to use this extension. In industrial projects, APIs grow quickly and if they happen to have this situation, they may have to use some other library if they don't want to duplicate the codebase of their API.

I've created another decorator `token_auth` that does token based authorization. But both `oidc_auth` & `token_auth` cannot be used together on the same view function. So it's either cookie based authentication or token based authorization which addresses your concern.

```
def token_auth(self, provider_name, scopes_required: list = None):
         '''Token based authorization.
 
         Parameters
         ----------
         provider_name : str
             Name of the provider registered with OIDCAuthorization.
         scopes_required : list, optional
             List of valid scopes associated with the endpoint.
 
         How To Use
         ----------
         >>> auth = OIDCAuthentication({'default': provider_config},
                                       access_token_required=True)
         >>> @app.route('/')
             @auth.token_auth(provider_name='default')
             def index():
                 ...
         >>> # You can also specify scopes required by your endpoint.
         >>> @auth.token_auth(provider_name='default',
                              scopes_required=['read', 'write'])
         '''
         def token_decorator(view_func):
 
             @functools.wraps(view_func)
             def wrapper(*args, **kwargs):
                 logger.debug('token executed')
                 client = self.clients[provider_name]
                 if not self._check_authorization_header(flask.request):
                     raise AccessDenied('missing authorization field')
                 token_introspection_result = self.introspect_token(
                     request=flask.request, client=client,
                     scopes=scopes_required)
                 if token_introspection_result:
                     logger.info('user has valid access token')
                     flask.session['token_introspection_info'] = token_introspection_result
                     return view_func(*args, **kwargs)
                 # Forbid access if the access token is invalid.
                 flask.abort(403)
             return wrapper
         return token_decorator
```
However, to address the problem with private API that can be made to accept both cookie and token, I've still kept the condition to check for token if cookie is not provided in `oidc_auth`. But this time it retrieves full information about the token and the user so overall it basically becomes decorator for authentication.

```
def oidc_auth(self, provider_name, accept_token: bool = None,
                   scopes_required: list = None):
         if provider_name not in self._provider_configurations:
             raise ValueError(
                 "Provider name '{}' not in configured providers: {}.".format(provider_name,
                                                                              self._provider_configurations.keys())
             )
         # If the user wants to enforce or disbale authorization on selective
         # endpoints, it can be enforced by providing boolean to accept_token
         # argument. If no boolean is explicitly provided, the value will be
         # taken from self.access_token_required which enforces authorization
         # globally and is False by default.
         if accept_token is None:
             accept_token = self.access_token_required
 
         def oidc_decorator(view_func):
             @functools.wraps(view_func)
             def wrapper(*args, **kwargs):
                 session = UserSession(flask.session, provider_name)
                 client = self.clients[session.current_provider]
 
                 if session.should_refresh(client.session_refresh_interval_seconds):
                     logger.debug('user auth will be refreshed "silently"')
                     return self._authenticate(client, interactive=False)
                 elif session.is_authenticated():
                     logger.debug('user is already authenticated')
                     return view_func(*args, **kwargs)
                 # If the user is not authenticated, check if the user has
                 # passed access token in the request header and make
                 # introspection request to the identity provider to verify the
                 # token.
                 elif accept_token and self._check_authorization_header(
                      flask.request):
                     token_introspection_result = self.introspect_token(
                         request=flask.request, client=client,
                         scopes=scopes_required)
                     if token_introspection_result:
                         logger.info('user has valid access token')
                         flask.session['token_introspection_info'] = token_introspection_result
                         return view_func(*args, **kwargs)
                     # Forbid access if the access token is invalid.
                     flask.abort(403)
                 # Else authenticate the user using Authorization Code Flow
                 else:
                     logger.debug('user not authenticated, start flow')
                     return self._authenticate(client)
 
             return wrapper
 
         return oidc_decorator
```
## What do you think?

---

### Current situation
flask_oidc extension is way outdated and [it's not complying with Oauth2.0 specs]( properly anymore so developers will likely adopt this extension if it takes little to no changes to their codebase. Based on the feedback that I've received from my team and community developers, these are the things they voted for:

1. A single decorator should be enough to protect APIs.
2. The integration with flask-pyoidc should require only minimal changes to the codebase.
3. It should be seamless as in it can turned on or off for selective endpoints by passing an argument.
4. A private API should be made accessible through both cookie based authentication and token based authorization.

Reply:
Sorry for my slow reply, holidays got in the way.
The `token_auth` method looks like what I was thinking with a separate decorator - thanks for sticking with it👍 .

I've realised yet one more reason for why I think fully splitting the 2 decorators though, i.e. not checking for any token i `oidc_auth` decorator, as it complicates the public API of this extension: the introspection response is stored under a separate key in the `flask.session` (`token_introspection_info`).
This means that if you're using the `oidc_auth` decorator with `accept_token=True` the user code still needs to do a slightly complicated check to see from which session key it should read the user data. This to me is a dealbreaker.

Also, I don't think the token introspection info should be stored in the `flask.session` at all.
If a token is used with a backend API (i.e. not related to any web frontend), I think it would be unexpected if it created a persistent session (returned as cookie).
It's a security risk, as it might lead to circumstances where a backend request is authenticated via cookies while it should only be authenticated via tokens.
Storing the introspection info in the request context (via `LocalProxy`) would be the way to go (see for example `current_user` from [flask-Login](https://github.com/maxcountryman/flask-login/blob/c760c0ef7ccc95d49b4693200245a4f2b148d41b/flask_login/utils.py#L25)).

Reply:
I've committed new code to address your feedback. `token_introspection_result` will now be saved in `flask.g` application context which will be available for current request only. Once the user does `flask.g.pop('token_introspection_result', None)` to read token metadata, it is cleared from the application context. Even if the user doesn't pop it, it will be overwritten with new token_introspection_info value on new request.

I've given more thoughts on how to apply oidc as well as token authorization on private APIs and current implementation is the only way that works with no changes to the user codebase. I'm in touch with a dev team who has analytical web app and they have this situation. Based on the feedback I've received from devs, they don't want to duplicate their API endpoints.

Ideally, it makes sense to keep `oidc_auth` and `token_auth` completely separate but to address all possible scenarios, can you make current implementation an exception for this?

In future, a better solution and based on the feedback of other contributors, we can make amends to it. What do you think?

Reply:
1. I was thinking the same of putting the introspection info in `flask.g`, but per the [Flask documentation](https://flask.palletsprojects.com/en/2.0.x/appcontext/#storing-data) it should be reserved for only user-code. Hence I think it would be better to create a separate LocalProxy variable to guarantee it won't overwrite any user data (however unlikely that is).
2. I've experimented a bit and with your new addition of `token_auth` decorator it actually is possible to combine the decorators to support both token and cookie/web auth:
    ```python
    def token_and_cookie_auth(provider_name, scopes_required):
        def combined_decorator(view_func):
            @functools.wraps(view_func)
            def wrapper(*args, **kwargs):
                try:
                    return auth.token_auth(provider_name, scopes_required)(view_func)(*args, **kwargs)
                except HTTPException as e:
                    if e.code == 401: # TODO this relies on tiny change in `token_auth` to call `flask.abort(401)` instead of `raise AccessDenied`
                        # no token in the request, try web login flow instead
                        return auth.oidc_auth(provider_name)(view_func)(*args, **kwargs)

                    raise # will just pass-through the 403 response if there was a token but it was invalid

            return wrapper

        return combined_decorator

    @app.route('/protected')
    @token_and_cookie_auth('default', ['scope1'])
    def protected():
        # TODO check for introspection info or user session
    ```
  
   Would this option work with your requirements?
  
   The reason why I'm so hesitant to mix the two together as the first implementation is that once it has been added it will be very hard to make amends to it (it will be a major API change for this extension). So I want to explore all options where the existing `oidc_auth` decorator can be kept entirely outside this use case 🙂.

Reply:
Earlier, I've used `_app_ctx_stack.top` as recommended by flask for extensions but our requirement is to allow user to obtain the information about the token so I used `flask.g` for this. If I use werkzeug's **LocalProxy**, will it be still possible for the user to obtain that information from application context?

Combining the decorators was a great solution. It precisely fits my requirement. I've created another decorator `access_control` based on your solution. The naming for this decorator I've taken from [this answer](https://security.stackexchange.com/a/109958). I've removed `_auth` suffix from its name because access_control is composed of authentication and authorization (too many auths here already). Naming suggestions are welcome.

```python
def access_control(self, provider_name: str,
                       scopes_required: list = None):
        '''This decorator serves dual purpose that is it can do both token
        based authorization and oidc based authentication. If your API needs
        to be accessible by either modes, use this decorator otherwise use
        either oidc_auth or token_auth.

        Parameters
        ----------
        provider_name : str
            Name of the provider registered with OIDCAuthorization.
        scopes_required : list, optional
            List of valid scopes associated with the endpoint.

        Raises
        ------
        HTTPException
            If the authorization is disabled.

        How To Use
        ----------
        >>> auth = OIDCAuthentication({'default': provider_config},
                                      disable_authorization=False)
        >>> @app.route('/')
            @auth.access_control(provider_name='default')
            def index():
                ...
        >>> # You can also specify scopes required by your endpoint.
        >>> @auth.access_control(provider_name='default',
                                 scopes_required=['read', 'write'])
        >>> # Additionally, you can disable token authorization entirely by
            # setting disable_authorization to True. It is helpful when you
            # want to deprecate developer access to your APIs.
        >>> auth = OIDCAuthentication({'default': provider_config},
                                      disable_authorization=True)
        '''
        def combined_decorator(view_func):

            fallback_to_oidc = self.oidc_auth(provider_name)(view_func)

            @functools.wraps(view_func)
            def wrapper(*args, **kwargs):

                try:
                    if self.disable_authorization:
                        logger.info('Authorization is disabled, this endpoint can use OIDC only')
                        flask.abort(401)
                    return self.token_auth(provider_name,
                                           scopes_required)(
                                               view_func)(*args, **kwargs)
                except HTTPException:
                    return fallback_to_oidc(*args, **kwargs)
                except AccessDenied:
                    return fallback_to_oidc(*args, **kwargs)
                except NotForMe:
                    return fallback_to_oidc(*args, **kwargs)

            return wrapper

        return combined_decorator
```

Now only `flask.g` is left to address once I receive your feedback. I'm also thinking to introduce token introspection for access_code saved in the cookie. If the user has cookie, we can have the access_code introspected along with checking the `last_authenticated` field.

Reply:
About the `LocalProxy`: I've not had time to test and implement it myself, but looking at for example how flask-login does it I assume it's available to user code as well (I understand that as the entire point of [`current_user`](https://github.com/maxcountryman/flask-login/blob/c760c0ef7ccc95d49b4693200245a4f2b148d41b/flask_login/utils.py#L25)).

Looking at your implementation of `access_control` decorator it seems you have some additional custom requirements, namely the `disable_authorization` flag.
It got me thinking it would be better to leave this to the extension users to write this "combinated" decorator.
That would allow full control of how/when which access control mechanism (token or cookie) should be applied - for example to deprecate, or differences in handling of 401/403 when introspecting a token.
Your implementation falls back in all cases, but it could be argued that if a token was included in the request but invalid that no fallback should happen.
I'm not saying either case is more correct, rather that this extension should probably not be opinionated about this and leave it up to users to decide the proper access control.
All this is to say I'm leaning towards this `access_control` decorator being better placed as an example in the documentation rather than included in the public API. 🤔 

>  I'm also thinking to introduce token introspection for access_code saved in the cookie

Are you referring to the access_token stored in the session? (There's no "access_code".)
If so, what is the introspection needed for?
All user data should already be in the session in that case.

Reply:
I've gone through werkzeug's **LocalProxy** and it has the same use case as flask `_app_ctx_stack.top`. They allow extensions to store internal data but it's not retrievable through **flask.g**. Our use case is to make `_token_introsoection_info` available to the user on application context. So if the extension wants to deliver that data to the user, flask.g is the way.

`disable_authorization` is not my own requirement but the reason of adding it is if your service is moving to version 2 APIs and you want to deprecate developer access to the version 1 APIs (maybe to encourage them to use version 2 API or maybe you only want the latest API version to have developer feature), then you can disable token based authorization for version 1 APIs with just one argument instead of making the user to revisit and modify the decorator. This is a lazy way but it's worth the time it saves to deprecate legacy API versions.

Asking users to implement hybrid functionality is not a good idea because users tend to write vulnerable code and security is hard. Besides, most developers prefer out of box support from extensions. I've received a feedback from a dev today and I'm told *"if the individual functionality is already offered by the extension, why can't it offer hybrid feature as well?"*.

I've committed new code which addresses when should the fallback to oidc happen. We should atleast offer the base hybrid functionality and if one isn't satisfied with the stock implementation, one can reimplement the hybrid decorator although it will be a rare usecase when one has to do it.

Reply:
With the LocalProxy it won't be accessible via `flask.g`.
Instead this extension would create a variable that can be imported in user code and would contain the introspection data if it exists.

Fair enough about the hybrid decorator, you make a good case for including it. 👍 

I'm not convinced about the `disable_authorization` flag though - the deprecation of an auth method in an API would be an edge case and not really something this extension needs to be concerned about.
I think the cost - one more configuration option that needs to be maintained and documented - is too high for something that can be achieved with a simple find-and-replace operation.

I've added some comments about exceptions and 401/403 response code.
But after that I think this is close to being ready. 👏 
Could you please add some tests that cover the new functionality?

Reply:
I've added `LocalProxy` for token introspection data. It can be accessed by using `auth.current_identity` in the view function. It is available for current request only and is destroyed after every request just like `flask.g`.